### PR TITLE
Eng/pr cas support next

### PR DESCRIPTION
**Pull request by cachemeifyoucan**

### CODE REVIEW

File: clang/lib/CodeGen/CodeGenModule.cpp [hunk 1]
```diff
@@ -6530,7 +6530,10 @@ void CodeGenFunction::EmitDeclMetadata() {
 void CodeGenModule::EmitVersionIdentMetadata() {
   llvm::NamedMDNode *IdentMetadata =
     TheModule.getOrInsertNamedMetadata("llvm.ident");
-  std::string Version = getClangFullVersion();
+
+  // FIXME: Choose IncludeRevision based on command-line argument. Including
+  // the Git revision is bad for reproducible builds / caching.
+  std::string Version = getClangFullVersion(/*IncludeRevision=*/false);
```
Comment (cachemeifyoucan) [inline clang/lib/CodeGen/CodeGenModule.cpp]: Do we need to upstream this?

File: clang/lib/Driver/ToolChains/Darwin.cpp [hunk 2]
```diff
@@ -759,9 +855,27 @@ void darwin::Linker::ConstructJob(Compilation &C, const JobAction &JA,
                        "-filelist"};
   }
 
+  // Set the environment to include ZERO_AR_FLAGS.
+  //
+  // FIXME: Add a flag to the linkers to support this without an environment
```
Comment (cachemeifyoucan) [inline clang/lib/Driver/ToolChains/Darwin.cpp:855]: Another reproducibility related upstreaming. Do we need this?

File: clang/tools/libclang/CDependencies.cpp [hunk 3]
```diff
@@ -53,8 +53,11 @@ void clang_experimental_ModuleDependencySet_dispose(
 
 CXDependencyScannerService
 clang_experimental_DependencyScannerService_create_v0(CXDependencyMode Format) {
+  // FIXME: Pass default CASOpts and nullptr as CachingOnDiskFileSystem now.
+  CASOptions CASOpts;
+  IntrusiveRefCntPtr<llvm::cas::CachingOnDiskFileSystem> FS;
   return wrap(new DependencyScanningService(
-      ScanningMode::DependencyDirectivesScan, unwrap(Format),
+      ScanningMode::DependencyDirectivesScan, unwrap(Format), CASOpts, FS,
```
Comment (cachemeifyoucan) [inline clang/tools/libclang/CDependencies.cpp:60]: The fix in this file is just get things compiling. I don't think we need them anymore

**Review by cachemeifyoucan — COMMENTED** [review]
@swift-ci please test

File: clang/lib/Driver/ToolChains/Darwin.cpp (hunk 2 above)
Comment (akyrtzi) [reply clang/lib/Driver/ToolChains/Darwin.cpp:855]: Seems useful.

File: clang/lib/CodeGen/CodeGenModule.cpp (hunk 1 above)
Comment (akyrtzi) [reply clang/lib/CodeGen/CodeGenModule.cpp]: It's not clear to me we need this, people (including us) should be living on releases, not a clang compiled locally that changes commit-to-commit.

**Comment by cachemeifyoucan** [comment]
@swift-ci Please smoke test

File: clang/include/clang/Frontend/CompilerInvocation.h [hunk 4]
```diff
@@ -242,6 +252,56 @@ class CompilerInvocation : public CompilerInvocationRefBase,
...
+                              SmallVectorImpl<const char *> &Args,
+                              CompilerInvocation::StringAllocator SA);
+  /// @}
+  /// @name Option Subgroups
+  /// @{
+
+  CASOptions &getCASOpts() { return CASOpts; }
+  const CASOptions &getCASOpts() const { return CASOpts; }
+
+  AnalyzerOptionsRef getAnalyzerOpts() const { return AnalyzerOpts; }
```
Comment (benlangmuir) [inline clang/include/clang/Frontend/CompilerInvocation.h]: Do you know where these extra accessors came from? We should already inherit all of them (except CASOpts) from `CompilerInvocationRefBase` or `CompilerInvocationValueBase`.

File: llvm/include/llvm/CAS/CASDB.h [hunk 5]
```diff
@@ -0,0 +1,528 @@
...
+  explicit LeafNodeProxy(NodeProxy N) : NodeProxy(std::move(N)) {
+    assert(this->NodeProxy::getNumReferences() == 0);
+  }
+};
+
+/// FIXME: Remove this after updating callers.
+using BlobProxy = LeafNodeProxy;
+
+Expected<std::unique_ptr<CASDB>>
+createPluginCAS(StringRef PluginPath, ArrayRef<std::string> PluginArgs = None);
```
Comment (benlangmuir) [inline llvm/include/llvm/CAS/CASDB.h]: This unused function and the (dummy) implementation in lib/CAS/PluginCAS.cpp should be removed.

File: llvm/include/llvm/Support/raw_ostream_proxy.h [hunk 6]
```diff
@@ -0,0 +1,158 @@
...
+      : RawOstreamAdaptorT(OS, std::forward<ArgsT>(Args)...) {}
+
+  raw_pwrite_stream &getProxiedOS() const {
+    return static_cast<raw_pwrite_stream &>(RawOstreamAdaptorT::getProxiedOS());
+  }
+};
+
+/// Non-owning proxy for a \a raw_ostream. Enables passing a stream into of an
+/// API that takes ownership.
+class raw_ostream_proxy : public raw_ostream_proxy_adaptor<> {
```
Comment (benlangmuir) [inline llvm/include/llvm/Support/raw_ostream_proxy.h:139]: Is `raw_ostream_proxy` used anywhere? I only see `raw_pwrite_stream_proxy` being used. Maybe we can drop this one?

File: llvm/include/llvm/Support/Path.h [hunk 7]
```diff
@@ -48,6 +48,14 @@ constexpr bool is_style_posix(Style S) {
 /// Check if \p S uses Windows path rules.
 constexpr bool is_style_windows(Style S) { return !is_style_posix(S); }
 
+constexpr Style system_style() {
```
Comment (benlangmuir) [inline llvm/include/llvm/Support/Path.h]: The only place this is used is to check if it's posix.  Instead we should check `is_style_posix(Style::native)` which already does the right thing, and lets us avoid modifying Path.h.

File: clang/tools/driver/clang-cache-build-session [hunk 8]
```diff
@@ -0,0 +1,9 @@
+#!/bin/sh
+
+# Set 'CLANG_CACHE_BUILD_SESSION_ID' to a unique identifier so that clang
+# invocations under the given command share the same depscan daemon while the
+# command is running.
+# Uses the process id to ensure parallel invocations of this script will not
+# share the same identifier, and 'elapsed seconds since epoch' to ensure the
+# same for consecutive invocations.
+env CLANG_CACHE_BUILD_SESSION_ID=$$-$(date +%s) "$@"
```
Comment (benlangmuir) [inline clang/tools/driver/clang-cache-build-session]: This is the old version before it became a real binary.  Maybe need to update the PR?

File: clang/tools/driver/driver.cpp [hunk 9]
```diff
@@ -50,6 +52,18 @@
...
+#ifndef _WIN32
+#include <unistd.h>
+#endif
+
+#if defined(__APPLE__) &&                                                      \
+    __has_include(<libproc.h>) && __has_include(<sys/proc_info.h>)
+#define USE_APPLE_LIBPROC_FOR_DEPSCAN_ANCESTORS
+#include <libproc.h>
+#include <sys/proc_info.h>
+#endif
```
Comment (benlangmuir) [inline clang/tools/driver/driver.cpp]: Are all these headers being used somewhere? I think this change may be dead.

File: clang/tools/driver/CMakeLists.txt [hunk 10]
```diff
@@ -82,6 +111,7 @@ if (APPLE)
   set(TOOL_INFO_PLIST_OUT "${CMAKE_CURRENT_BINARY_DIR}/${TOOL_INFO_PLIST}")
   target_link_libraries(clang
     PRIVATE
+    -lutil # FIXME: Apple-only. Not portable
```
Comment (benlangmuir) [inline clang/tools/driver/CMakeLists.txt]: What are we using from this library?

**Comment by cachemeifyoucan** [comment]
Update

**Comment by cachemeifyoucan** [comment]
@swift-ci please test

**Comment by cachemeifyoucan** [comment]
@swift-ci please test

**Comment by cachemeifyoucan** [comment]
@swift-ci please test

**Pull request merged**